lib/route/cls/u32.c

```diff
@@ -197,6 +197,24 @@ static int u32_clone(void *_dst, void *_src)
 			return -NLE_NOMEM;
 
 		memcpy(dst->cu_act, src->cu_act, sizeof(struct rtnl_act));
+
+		/* action nl list next and prev pointers must be updated */
+		nl_init_list_head(&dst->cu_act->ce_list);
+
+		if (src->cu_act->c_opts && !(dst->cu_act->c_opts = nl_data_clone(src->cu_act->c_opts)))
+			return -NLE_NOMEM;
+
+		if (src->cu_act->c_xstats && !(dst->cu_act->c_xstats = nl_data_clone(src->cu_act->c_xstats)))
+			return -NLE_NOMEM;
+
+		if (src->cu_act->c_subdata && !(dst->cu_act->c_subdata = nl_data_clone(src->cu_act->c_subdata)))
+			return -NLE_NOMEM;
```

Comment:
all these return statement still leave the instance in an invalid state. That must be properly handled.

I would suggest to use the new `_nl_auto` macros and first clone the fields all (assigning them to `__attribute__((cleanup))` variables  on the stack), before transferring ownership at once (where we know it's no longer going to fail).

Reply:
I see your point. I've reworked this function now. Please have a look when you've got time.